mitmproxy/websocket.py

```diff
@@ -1,168 +1,126 @@
-    content: Union[bytes, str]
+    type: Opcode
+    """
+    The message type, as per RFC 6455's [opcode](https://tools.ietf.org/html/rfc6455#section-5.2).
+
+    Note that mitmproxy will always store the message contents as *bytes*.
+    A dedicated `.text` property for text messages is planned, see https://github.com/mitmproxy/mitmproxy/pull/4486.
+    """
+    content: bytes
     """A byte-string representing the content of this message."""
     timestamp: float
     """Timestamp of when this message was received or created."""
-
     killed: bool
-    """True if this messages was killed and should not be sent to the other endpoint."""
+    """True if the message has not been forwarded by mitmproxy, False otherwise."""
 
     def __init__(
         self,
-        type: int,
+        type: Union[int, Opcode],
         from_client: bool,
-        content: Union[bytes, str],
+        content: bytes,
         timestamp: Optional[float] = None,
-        killed: bool = False
+        killed: bool = False,
     ) -> None:
-        self.type = Opcode(type)  # type: ignore
         self.from_client = from_client
+        self.type = Opcode(type)
         self.content = content
         self.timestamp: float = timestamp or time.time()
         self.killed = killed
 
     @classmethod
-    def from_state(cls, state):
+    def from_state(cls, state: WebSocketMessageState):
         return cls(*state)
 
-    def get_state(self):
+    def get_state(self) -> WebSocketMessageState:
         return int(self.type), self.from_client, self.content, self.timestamp, self.killed
 
-    def set_state(self, state):
-        self.type, self.from_client, self.content, self.timestamp, self.killed = state
-        self.type = Opcode(self.type)  # replace enum with bare int
+    def set_state(self, state: WebSocketMessageState) -> None:
+        typ, self.from_client, self.content, self.timestamp, self.killed = state
+        self.type = Opcode(typ)
 
     def __repr__(self):
         if self.type == Opcode.TEXT:
-            return "text message: {}".format(repr(self.content))
+            return repr(self.content.decode(errors="replace"))
         else:
-            return "binary message: {}".format(strutils.bytes_to_escaped_str(self.content))
+            return repr(self.content)
+
+    def kill(self):
+        # Likely to be replaced with .drop() in the future, see https://github.com/mitmproxy/mitmproxy/pull/4486
+        self.killed = True
 
-    def kill(self):  # pragma: no cover
-        """
-        Kill this message.
 
-        It will not be sent to the other endpoint. This has no effect in streaming mode.
-        """
-        warnings.warn(
-            "WebSocketMessage.kill is deprecated, set an empty content instead.",
-            DeprecationWarning,
-            stacklevel=2,
-        )
-        # empty str or empty bytes.
-        self.content = type(self.content)()
+class WebSocketData(stateobject.StateObject):
```

Comment:
What do you think about `WebSocketDetails`? `data` has a certain connotation

Reply:
I picked "Data" over "Details" as I felt it's really not just the details, it's also the main part. 😛 
What connotation do you have for `data`?

Other alternatives:

 - WebSocketConnection: Bad idea IMO, that clashes with `mitmproxy.connection`.
 - WebSocketFlow: Bad idea IMO, this is not a top-level flow.
 - WebSocketPart
 - WebSocketExchange
 - WebSocketInfo
 - WebSocketMessages

Nothing really convinces me, I feel that "data" as in dataclass is the most straightforward.

Reply:
`WebSocketData` sounds great :+1: :smile: